library/alloc/src/boxed.rs

```diff
@@ -578,6 +578,62 @@ impl<T, A: Allocator> Box<T, A> {
     {
         *boxed
     }
+
+    /// Safely reads the value on heap without deallocating.
+    ///
+    /// This method is mostly useful for avoiding reallocations.
+    /// The method consumes the box to prevent double-free.
+    /// The value on heap should be considered truly uninitialized and MUST NOT be read.
+    /// Specifically, this code is definitely UB:
+    ///
+    /// ```no_run
+    /// #![feature(new_uninit)]
+    ///
+    /// let value = Box::new("hello".to_owned());
+    /// // Double free here!
+    /// unsafe { Box::take(value).1.assume_init(); }
+    /// ```
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(new_uninit)]
+    ///
+    /// let boxed_value = Box::new("hello".to_owned());
+    /// let (mut value, allocation) = Box::take(boxed_value);
+    /// assert_eq!(value, "hello");
+    /// // More realistic code would consume the value and produce other value.
+    /// // For simple demonstration we just modify it here.
+    /// value.push_str(" world");
+    /// let boxed_value = Box::write(allocation, value);
+    /// assert_eq!(*boxed_value, "hello world");
+    /// ```
+    #[unstable(feature = "new_uninit", issue = "63291")]
+    #[rustc_const_unstable(feature = "const_box", issue = "92521")]
+    #[inline]
+    pub const fn take(this: Self) -> (T, Box<mem::MaybeUninit<T>, A>) {
```

Comment:
`take` to me has a very strong implication of `mem::take` (or `Option::take` or ...) to me, with `&mut self` and such, so I would encourage a different name here.

👍 to the functionality, though!  Makes good sense to me.

Reply:
I actually picked the name because it is similar to `mem::take` in the sense of "taking the value out and replacing it with something", in this case replacing it with nothing and tracking it in the state. However if people feel it's strongly associated with `&mut self` I'm open to a better name, I'd just like to hear some suggestions on alternatives or a feedback on my own ides for alternatives (see above).

Reply:
I do see the parallel, though to me this API doesn't do the "and replacing it with something" part that distinguishes `take` from `into_inner` kinds of things.

But more importantly to me, `mem::take(&mut my_box)` works today, so I think `Box::take` doing something pretty different would be surprising.  If anything, I might expect it to be `mem::take(&mut *my_box)` -- note the deref -- like how `RefCell::take` works.  (And note that `mem::take(&mut my_option)` does exactly the same thing as `Option::take(&mut my_option)`, as another example.)

Reply:
Yeah, what do you think bout `into_parts()`? That one feels like the next best to me now.